File: src/main/java/com/atsuishio/superbwarfare/init/ModRecipes.java
package com.atsuishio.superbwarfare.init;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.recipe.*;
import com.atsuishio.superbwarfare.recipe.vehicle.VehicleAssemblingRecipe;
import com.atsuishio.superbwarfare.recipe.vehicle.VehicleAssemblingRecipeSerializer;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.item.crafting.RecipeType;
import net.minecraft.world.item.crafting.SimpleCraftingRecipeSerializer;

import java.util.function.Supplier;

public class ModRecipes {

    public static final Supplier<RecipeSerializer<PotionMortarShellRecipe>> POTION_MORTAR_SHELL_SERIALIZER =
            Registration.recipeSerializer("potion_mortar_shell", () -> new SimpleCraftingRecipeSerializer<>(PotionMortarShellRecipe::new));
    public static final Supplier<RecipeSerializer<AmmoBoxAddAmmoRecipe>> AMMO_BOX_ADD_AMMO_SERIALIZER =
            Registration.recipeSerializer("ammo_box_add_ammo", () -> new SimpleCraftingRecipeSerializer<>(AmmoBoxAddAmmoRecipe::new));
    public static final Supplier<RecipeSerializer<AmmoBoxExtractAmmoRecipe>> AMMO_BOX_EXTRACT_AMMO_SERIALIZER =
            Registration.recipeSerializer("ammo_box_extract_ammo", () -> new SimpleCraftingRecipeSerializer<>(AmmoBoxExtractAmmoRecipe::new));
    public static final Supplier<RecipeSerializer<SmokeDyeRecipe>> SMOKE_DYE_SERIALIZER =
            Registration.recipeSerializer("smoke_dye", () -> new SimpleCraftingRecipeSerializer<>(SmokeDyeRecipe::new));
    public static final Supplier<RecipeSerializer<VehicleAssemblingRecipe>> VEHICLE_ASSEMBLING_SERIALIZER =
            Registration.recipeSerializer("vehicle_assembling", VehicleAssemblingRecipeSerializer::new);
    public static final Supplier<RecipeSerializer<VehicleResetRecipe>> VEHICLE_RESET_SERIALIZER =
            Registration.recipeSerializer("vehicle_reset", () -> new SimpleCraftingRecipeSerializer<>(VehicleResetRecipe::new));
    public static final Supplier<RecipeSerializer<CupidArrowRecipe>> CUPID_ARROW_SERIALIZER =
            Registration.recipeSerializer("cupid_arrow", () -> new SimpleCraftingRecipeSerializer<>(CupidArrowRecipe::new));
    public static final Supplier<RecipeSerializer<C4BombRcRecipe>> C4_BOMB_RC_SERIALIZER =
            Registration.recipeSerializer("c4_bomb_rc", () -> new SimpleCraftingRecipeSerializer<>(C4BombRcRecipe::new));

    public static final Supplier<RecipeType<VehicleAssemblingRecipe>> VEHICLE_ASSEMBLING_TYPE =
            Registration.recipeType("vehicle_assembling", () -> new RecipeType<>() {
                @Override
                public String toString() {
                    return Mod.MODID + ":vehicle_assembling";
                }
            });

    public static void register() {

    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/init/ModRecipes.kt
package com.atsuishio.superbwarfare.init

import com.atsuishio.superbwarfare.Mod
import com.atsuishio.superbwarfare.recipe.*
import com.atsuishio.superbwarfare.recipe.vehicle.VehicleAssemblingRecipe
import com.atsuishio.superbwarfare.recipe.vehicle.VehicleAssemblingRecipeSerializer
import net.minecraft.world.item.crafting.RecipeSerializer
import net.minecraft.world.item.crafting.RecipeType
import net.minecraft.world.item.crafting.SimpleCraftingRecipeSerializer
import java.util.function.Supplier

@Suppress("unused")
object ModRecipes {
    @JvmField
    val CUPID_ARROW_SERIALIZER: RecipeSerializer<CupidArrowRecipe> =
        Registration.recipeSerializer("cupid_arrow") {
            SimpleCraftingRecipeSerializer { id, category -> CupidArrowRecipe(id, category) }
        }

    @JvmField
    val C4_BOMB_RC_SERIALIZER: RecipeSerializer<C4BombRcRecipe> =
        Registration.recipeSerializer("c4_bomb_rc") {
            SimpleCraftingRecipeSerializer { id, category -> C4BombRcRecipe(id, category) }
        }

    @JvmField
    val POTION_MORTAR_SHELL_SERIALIZER: RecipeSerializer<PotionMortarShellRecipe> =
        Registration.recipeSerializer("potion_mortar_shell") {
            SimpleCraftingRecipeSerializer { id, category -> PotionMortarShellRecipe(id, category) }
        }

    @JvmField
    val SMOKE_DYE_SERIALIZER: RecipeSerializer<SmokeDyeRecipe> =
        Registration.recipeSerializer("smoke_dye") {
            SimpleCraftingRecipeSerializer { id, category -> SmokeDyeRecipe(id, category) }
        }

    @JvmField
    val VEHICLE_ASSEMBLING_SERIALIZER: RecipeSerializer<VehicleAssemblingRecipe> =
        Registration.recipeSerializer("vehicle_assembling") { VehicleAssemblingRecipeSerializer() }

    @JvmField
    val VEHICLE_RESET_SERIALIZER: RecipeSerializer<VehicleResetRecipe> =
        Registration.recipeSerializer("vehicle_reset") {
            SimpleCraftingRecipeSerializer { id, category -> VehicleResetRecipe(id, category) }
        }

    @JvmField
    val RESEARCHING_SERIALIZER: RecipeSerializer<ResearchingRecipe> =
        Registration.recipeSerializer("researching") { ResearchingRecipe.Serializer() }

    @JvmField
    val VEHICLE_ASSEMBLING_TYPE: RecipeType<VehicleAssemblingRecipe> =
        Registration.recipeType("vehicle_assembling") {
            object : RecipeType<VehicleAssemblingRecipe> {
                override fun toString(): String {
                    return Mod.MODID + ":vehicle_assembling"
                }
            }
        }

    @JvmField
    val RESEARCHING_TYPE: RecipeType<ResearchingRecipe> =
        Registration.recipeType("researching") {
            object : RecipeType<ResearchingRecipe> {
                override fun toString(): String {
                    return Mod.MODID + ":researching"
                }
            }
        }

    @JvmStatic
    fun init() {
    }
}
